### ISSUE TITLE
Recommended structure for external protobufs

### ISSUE THREAD

**Issue by trystanj** [bazelbuild/rules_go#877]
I would like to keep `.proto` files in a separate, private repository, shared by multiple projects in different languages. In my case, the protos I'm interested are in a folder in the remote repo.

What is the correct solution for this type of use-case? I worked through proto building when the protos were local and everything works great. If I want to move to a separate (private) repository, I haven't been able to figure out what needs to change.

Do I keep the BUILD.bazel files in the external repo? I assume it's no longer a `go_repository`, and since it's private it seems I will need to use Skylark's git rules so it uses system git (source: https://docs.bazel.build/versions/master/be/workspace.html#git_repository). How do I configure my local WORKSPACE correctly?

I know the protobuf support is still in the process of being reworked, so forgive me if I'm jumping the gun!

**Comment by jayconrod** [comment]
Sorry for not responding sooner. Got caught up with GothamGo.

cc @ianthehat for a definitive answer. I think you should put BUILD.bazel with with `proto_library` rules in that separate repository. You may or may not want to add `go_proto_library` rules to that repository as well, depending on whether they would be shared by other repos.

```
load("@io_bazel_rules_go//proto:def.bzl", "go_proto_library")

proto_library(
    name = "foo_proto",
    srcs = ["foo.proto"],
    visibility = ["//visibility:public"],
)

go_proto_library(
    name = "foo_go_proto",
    importpath = "example.com/proto",
    proto = ":foo_proto",
    visibility = ["//visibility:public"],
)
```

You can depend on that with `git_repository` in your WORKSPACE.

```
git_repository(
    name = "com_example_proto",
    remote = "https://example.com/proto",
    commit = "abcdef",
)
```

And then you can reference the libraries like this:

```
go_library(
    name = "go_default_library",
    srcs = ["bar.go"],
    deps = ["@com_example_proto//:foo_go_proto"],
    ...
)
```

Does that answer your question?

**Comment by trystanj** [comment]
No worries, I know I'm kind of getting ahead of the project / documentation.

I think I have more or less what you suggest, but I'm getting an error building. Here's my current structure for everything. I hope this provides some context - perhaps the structure doesn't make sense, in which case I'm happy to change it to conform to best practices! The shared protos will be used by multiple languages, FYI. I don't know if that changes anything or not.

### Main repo @ github.com/company/repo (private repository)
**WORKSPACE:**
```
...
git_repository(
    name = "com_github_company_proto",
    commit = "abcdef",
    remote = "github.com/company/proto",
)
...
```

**github.com/company/repo/project/project1/BUILD.bazel:**
```
...
go_library(
    name = "go_default_library",
    srcs = ["file1.go"],
    importpath = "github.com/company/repo/project/project1",
    visibility = ["//visibility:private"],
    deps = [
        ....
        "@com_github_company_proto//protogroup:go_default_library",
        ...
    ],
)
...
```

**github.com/company/repo/project/project1/file.go**
```
...
import protos "github.com/company/proto/protogroup"
....
```

-----

### Proto repo @ github.com/company/proto (also private repository)
**Structure:**
```
README.md
/protogroup
    BUILD.bazel
    proto1.proto
    proto2.proto
    ....
```

**BUILD.bazel**
```
load("@io_bazel_rules_go//go:def.bzl", "go_library")
load("@io_bazel_rules_go//proto:def.bzl", "go_proto_library")

proto_library(
    name = "protogroup_proto",
    srcs = [
      "proto1.proto",
      "proto2.proto",
      ],
    deps = [
        "@com_google_protobuf//:timestamp_proto",
    ],
    visibility = ["//visibility:public"],
)

go_proto_library(
    name = "go_default_library",
    proto = ":protogroup_proto",
    importpath = "github.com/company/proto/protogroup",
    deps = [
        "@com_github_golang_protobuf//ptypes/timestamp:go_default_library",
    ],
    visibility = ["//visibility:public"],
)
```

-----

### Bazel Build Error
```
➜  repo git:(use-repo) ✗ bazel build //...
ERROR: /Users/me/go/src/github.com/company/repo/project/project1/BUILD.bazel:3:1: no such package '@com_github_company_proto//protogroup': Invalid Git repository URI: Invalid remote: origin and referenced by '//project/project1/file:go_default_library'.
ERROR: Analysis of target '//project/project1/file:go_default_library' failed; build aborted.
INFO: Elapsed time: 20.184s
```

**Comment by jayconrod** [comment]
I think you're getting this error because you're missing `https://` in your `git_repository` remote. The string that's provided there has to be something that `git clone` would accept.

Note that this is different than `go_repository`, which does not require a scheme to be part of `importpath`. It will automatically determine that (and follow redirects, etc).

One other thing to point out: `http_archive` is a faster alternative to `git_repository`, but it requires a stable archive that you can fetch. It's good if you depend on a released version. Be careful not to use GitHub's automatically generated archives though, since they're generated on demand and don't have a stable sha256 sum.

**Comment by trystanj** [comment]
Ah! That solved that problem, thank you!

I'm running into issues compiling the protobufs now. Are the import paths intended to be absolute or relative?

Say `proto1` imports `proto2` above, and they are intended to be the same package (`protogroup`).

Should proto1 look like

1. `import "github.com/company/proto/protogroup/proto2";`
1. `import "proto2";`

I've tried both, and every combination in between, and keep having build errors.

```
➜  repo git:(use-repo) ✗ bazel build //...
INFO: Found 24 targets...
ERROR: /private/var/tmp/_bazel_trystanj/<commit>/external/com_github_company_proto/protogroup/BUILD.bazel:4:1: Generating Descriptor Set proto_library @com_github_company_proto//protogroup:protogroup_proto failed (Exit 1).
proto/protogroup/proto1.proto: File not found.
proto/protogroup/proto2.proto: File not found.
protogroup/proto3.proto: Import "proto/protogroup/proto1.proto" was not found or had errors.
protogroup/proto3.proto: Import "proto/protogroup/proto2.proto" was not found or had errors.
protogroup/proto3.proto:12:9: "Proto1" is not defined.
protogroup/proto3.proto:13:9: "Proto2" is not defined.
```

**Comment by jayconrod** [comment]
For importing one .proto from another .proto file, you need to give the path relative to the root of your repository. From your error message, it looks like the `proto_library` rule is in the directory `protogroup`, so assuming your .proto files are in the same directory, they would import each other as below. Make sure not to leave off the .proto extension.

```
import "protogroup/proto2.proto";
```

There isn't any way to import protos across repositories right now, but since all your protos are in the same repository, that shouldn't be a problem.

Go code should import proto libraries according to their import path, `github.com/company/proto/protogroup` in your example. It looks like you already have that working.

**Comment by trystanj** [comment]
Yeah, so that's what I've got set now, and I can compile them independently with `protoc`.

Now I'm seeing this error:
```
INFO: Found 1 target...
ERROR: /private/var/tmp/_bazel_trystanj/652a854d748ed815e21eca679201b497/external/com_github_company_proto/protogroup/BUILD.bazel:18:1: Generating into bazel-out/darwin_x86_64-fastbuild/bin/external/com_github_company_proto/protogroup/github.com/company/proto/protogroup failed (Exit 1).
../com_github_company_proto/protogroup/proto1.proto: No such file or directory
```

I think the compiler is looking in the wrong place; perhaps there's a proto rule setting I'm missing. I'm going to create two public repos that I can push to repro this instead, if that helps.

**Comment by trystanj** [comment]
@jayconrod I've strung together a minimal example. Bazel 0.5.4.

https://github.com/trystanj/external-proto & https://github.com/trystanj/proto

```
➜  external-proto git:(master) ✗ bazel build //...
INFO: Found 4 targets...
ERROR: /private/var/tmp/_bazel_trystanjohnson/2606cb74475ebda56a3b753b112a59a0/external/com_github_trystanj_proto/group/BUILD.bazel:18:1: Generating into bazel-out/darwin_x86_64-fastbuild/bin/external/com_github_trystanj_proto/group/github.com/trystanj/proto/group failed (Exit 1).
../com_github_trystanj_proto/group/outer.proto: No such file or directory
INFO: Elapsed time: 25.133s, Critical Path: 0.05s
```

**Comment by trystanj** [comment]
Okay, your recommended structure was correct, and it works! I was falling prey to #909, which was fixed by #911. Upgrading my `rules_go` SHA resolved the issue.

I'll work on upgrading to Bazel 0.6.0, but it seems (so far) we don't use rules that conflict, so a straight `rules_go` upgrade works.

Closing this issue - thank you for your help!

**Comment by CareF** [comment]
This doesn't seems to work for gRPC python
